### PR TITLE
Preserve order of visiting subexpressions in PatternVariableFinder

### LINKED ISSUE #9258: Please preserve order of subexpression visiting in PatternVariableFinder
See https://github.com/dotnet/roslyn/pull/9216/files#r54207731

/cc @AlekseyTs

### PR DESCRIPTION
**Pull request by gafter**
Fixes #9258

Previously reviewed as #9329

### CODE REVIEW

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Can I please have a second review on this?

File: src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs [hunk 1]
```diff
@@ -85,17 +70,30 @@ public override void VisitQueryExpression(QueryExpressionSyntax node)
...
+
+            var operands = ArrayBuilder<ExpressionSyntax>.GetInstance();
+            ExpressionSyntax current = node;
+            do
+            {
+                var binOp = (BinaryExpressionSyntax)current;
+                operands.Push(binOp.Right);
+                current = binOp.Left;
+            }
+            while (current is BinaryExpressionSyntax);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: Is this to handle deeply nested patterns? Do we expect such cases? And are we testing this?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: No, the code to _find whether or not there are patterns in a given expression_ did not handle deep expressions well, even when there are no patterns present.

**Comment by cston** [comment]
LGTM

**Pull request merged**